packages/clay-color-picker/src/Custom.tsx

```diff
@@ -0,0 +1,249 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+import GradientSelector from './GradientSelector';
+import Hue from './Hue';
+import React, {useState} from 'react';
+import Splotch from './Splotch';
+import tinycolor from 'tinycolor2';
+import {useHexInput} from './hooks';
+
+/**
+ * Renders custom color icon
+ * @return {React.Component}
+ */
+const CustomColorIcon: React.SFC = () => {
+	return (
+		<svg
```

Comment:
We can use the lexicon icons directly without having to add the svg hard code. Maybe we can create an Icon component soon, since we can use it a lot.

See example https://github.com/liferay/clay/pull/1440/files?file-filters%5B%5D=.tsx#diff-4fb77ee218c192d6a19607263ff8d304

Reply:
How do you feel about the pattern of having to pass `spriteMap` into the top level component? I saw you did it here, https://github.com/liferay/clay/pull/1440/files?file-filters%5B%5D=.tsx#diff-f83ebf28c6e7caeca167ae49475dcec9R20. Seems a bit clunky to me, since the the intent of the component is to be based on clay-css's icons. If we follow this pattern, every component that uses a clay icon, we have to provide a `spriteMap` prop at the top level component.

I was thinking in `clay-icon` we can reference clay-css's icon sprite sheet directly, that way we only have to reference the symbol. Then once `clay-icon` is merged(I'm working on it now), we can update this hardcoded value for that component.

Thoughts?

Reply:
- reference to clay-css: The concern about this was that there could be cases of loading more than one `icons.svg` in different parts of the application and in the portal and the network overhead it can have, can make the pages slow, this file is a bit large and inside the portal we are loading it.
- href hard code: we had these concerns before about adding the spritemap hard code, the concern was that in the portal the address may be different and we must give freedom to modify this URL. Because of this we had to add `spritemap` to all our components.

I think we can try to minimize that impact a bit, ClayIcon may have support for getting the `context` spritemap for full React applications, you can add the `provider` to your root component to pass the spritemap information, so you do not have to go through for all components.

Reply:
Oh good idea with context. I'll submit a PR for icon today to get some feedback on that idea. Gonna see about allowing either a prop or context

Reply:
In a way, that's how @wincent and @julien ended up doing [some of it in AlloyEditor](https://github.com/liferay/alloy-editor/blob/master/src/components/buttons/button-icon.jsx#L16).

One thing that's wrong there, and might need to rethink here as well is the `href` on svgs. SVGs are a basic static resource to host on a CDN. Unfortunately, svg does not support CORS by spec (at least last time I checked), so the following:

```
<svg href="cdn.foo.com/icons.svg">...</svg>
```

Will throw a CORS error when accessed from a different domain (`main.foo.com`). To still be able to load the svg spritemap, we're using [svg4everybody](https://github.com/jonathantneal/svg4everybody), which will use AJAX (granted the server is properly configured) to overcome this limitation. That, however, won't save us from the initial browser-triggered request, so the final change we might need to do is to use `data-href` instead of `href`.

At least, those are the things that we need to consider... another option would be to fin an alternative way to do this :)

Reply:
For additional reference:
- https://bugs.chromium.org/p/chromium/issues/detail?id=470601
- https://www.w3.org/Graphics/SVG/WG/track/actions/3781

Reply:
@matuzalemsteles check out https://github.com/liferay/clay/pull/1725/files for clay-icon with context. You can look at the demo file to see how it is used.

Reply:
More references:
- https://issues.liferay.com/browse/LPS-71763
- https://issues.liferay.com/browse/LPS-72903

Please, keep these in mind so we don't regress there

Reply:
thanks @jbalsas for context, one problem with using `href` or `data-href` (I'm not sure about `date-href`) is that it does not work in Safari 😞.

Thanks @bryceosterhaus I'm going to take a look, just a few concerns I have now, is that maybe it's more a context for the application, some applications may have many contexts aligned, I do not really like it, but if Clay comes to use context, maybe we can create a context for Clay and use it on every component you need. Something like `ClayContext`. Just to not have a connection for each component.

Reply:
I meant with the `<use>` tag. And the `data-href` needs to be used in conjunction with something like `svg4everybody` properly configured to fetch from `data-href` and avoid the browser initiating a request.

Another option we explored was to "force" adding the spritemap in the page and then just refer to the ids locally

Reply:
Oh okay, makes sense now 🙂

Reply:
@matuzalemsteles 
>  some applications may have many contexts aligned, I do not really like it, but if Clay comes to use context, maybe we can create a context for Clay and use it on every component you need. Something like ClayContext. Just to not have a connection for each component.

Even if an app uses many contexts, this shouldn't be too big of a problem since you can create an global context that consumes the other contexts. This could be something we could try for a global ClayContext. Each component can independently create their own, but also we can join them all together in a separate util that provides a global one.

Also, by allowing for context here, its just another feature and they can always use the spriteMap prop if they prefer.

Reply:
Good idea, I had not thought about it.

Reply:
Long thread, but yes, this kind of pattern (having to pass something down all the way from the top to a deeply nested component that renders an icon and is used everywhere) is an ideal use case for context. We should definitely explore that.